palm/containerizer.py

```diff
@@ -0,0 +1,184 @@
+
+        Args:
+            ctx (click.context): The click context object from the calling command
+            template_dir (Path): Path to the templates directory
+        """        
+        self.ctx = ctx
+        self.project_name = ctx.obj.palm.image_name
+        self.template_dir = template_dir
+        self.python_version = python_version
+
+    def run(self) -> None:
+        """Run the containerizer"""
+        if not self.validate_python_version():
+            click.secho(f"Invalid python version: {self.python_version}", fg="red")
+            return
+
+        try:
+            self.validate_python_version()
+            super().check_setup()
+        except AbortPalm as e:
+            click.secho(str(e), fg="red")
+            return
+
+        package_manager = self.package_manager()
+
+        if package_manager == "unknown":
+            try:
+                self.optionally_add_requirements_txt()
+            except AbortPalm:
+                click.secho("Aborting containerization", fg="red")
+                return
+            package_manager = "pip3"
+
+        target_dir = Path.cwd()
+        replacements = {
+            "project_name": self.project_name,
+            "package_manager": package_manager,
+            "python_version": self.python_version
+        }
+
+        super().generate(target_dir, replacements)
+
+    def package_manager(self) -> str:
+        """Determine which package manager is in use,
+        supports pip3 and poetry, returns "unknown" if neither are found
+
+        Returns:
+            str: pip3 | poetry | unknown
+        """        
+        if self.has_requirements_txt():
+            return "pip3"
+        if self.has_poetry():
+            return "poetry"
+
+        return "unknown"
+
+    def has_requirements_txt(self) -> bool:
+        """Check for a requirements.txt file in the project root"""
+        return Path("requirements.txt").exists()
+
+    def has_poetry(self) -> bool:
+        """Check for a poetry.lock file in the project root"""
+        return Path("poetry.lock").exists()
+
+    def optionally_add_requirements_txt(self):
+        """Optionally, add a requirements.txt file to the project root if it doesn't exist
+
+        Raises:
+            AbortPalm: Abort if user does not want to add requirements.txt
+        """        
+        use_pip = click.confirm(
+            "Unable to detect python package manger, requirements.txt will be used by default. Continue?"
+        )
+        if use_pip:
+            Path("requirements.txt").touch()
+        else:
+            raise AbortPalm("Aborting")
+
+    def validate_python_version(self) -> bool:
+        """Validate the python version provided by the user
+
+        Args:
+            python_version (str): The python version to validate
+
+        Returns:
+            bool: True if the python version is valid
+        """
+        match = re.match(r"^3\.[0-9]{1,2}$", self.python_version)
```

Comment:
we _could_ support 2.7 containers if we wanted to, right? that might be useful for a lot of legacy stuff

Reply:
I'm not sure, probably... I was intentionally being quite cautious with this but we probably don't need to worry _too_ much about what version is running inside the container, so much as what version is running on the metal.